### PR TITLE
Use derive attributes for uninteresting traversals

### PR DESCRIPTION
**Pull request by eggyal**
It appears that visiting and folding was implemented on `BitMatrix` solely so that the derive macros could be used on `GeneratorLayout`, however such implementation would not necessarily be correct for other uses (if there were any).  Adding attributes to the derive macro is more correct and potentially more generally useful.

r? @oli-obk

### CODE REVIEW

File: compiler/rustc_macros/src/type_foldable.rs [hunk 1]
```diff
@@ -16,8 +16,29 @@ pub fn type_foldable_derive(mut s: synstructure::Structure<'_>) -> proc_macro2::
         let bindings = vi.bindings();
         vi.construct(|_, index| {
             let bind = &bindings[index];
-            quote! {
-                ::rustc_middle::ty::fold::ir::TypeFoldable::try_fold_with(#bind, __folder)?
+
+            // fix fields with #[type_foldable(fixed)]
```
Comment (oli-obk) [inline compiler/rustc_macros/src/type_foldable.rs:20]: the meaning of `fix` or `fixed` is not clear to me from just the name. Maybe `identity` would be better?

**Review by oli-obk — CHANGES_REQUESTED** [review]
These attributes should be documented. On the trait maybe? not sure where we document the derives.

File: compiler/rustc_macros/src/type_foldable.rs (hunk 1 above)
Comment (eggyal) [reply compiler/rustc_macros/src/type_foldable.rs:20]: Yes, fair point.  Originally I considered `invariant` and `self`, didn't think either was clear enough so came up with `fixpoint` but thought it was a bit wordy.  I agree `identity` is probably best.

**Comment by eggyal** [comment]
We can document the macro.

**Comment by oli-obk** [comment]
Great thanks!

@bors r+ rollup

**Comment by eggyal** [comment]
Apologies, it failed tidy check.  Have re-pushed.
